java/src/main/java/ai/rapids/cudf/MultiBufferDataSource.java

```diff
@@ -0,0 +1,237 @@
+        startOffsets = new long[numBuffers];
+
+        long currentOffset = 0;
+        for (int i = 0; i < numBuffers; i++) {
+            HostMemoryBuffer hmb = buffers[i];
+            hmb.incRefCount();
+            hostBuffers[i] = hmb;
+            startOffsets[i] = currentOffset;
+            currentOffset += hmb.getLength();
+        }
+        sizeInBytes = currentOffset;
+        allocator = DefaultHostMemoryAllocator.get();
+    }
+
+    public MultiBufferDataSource(HostMemoryAllocator allocator, HostMemoryBuffer ... buffers) {
+        int numBuffers = buffers.length;
+        hostBuffers = new HostMemoryBuffer[numBuffers];
+        startOffsets = new long[numBuffers];
+
+        long currentOffset = 0;
+        for (int i = 0; i < numBuffers; i++) {
+            HostMemoryBuffer hmb = buffers[i];
+            hmb.incRefCount();
+            hostBuffers[i] = hmb;
+            startOffsets[i] = currentOffset;
+            currentOffset += hmb.getLength();
+        }
+        sizeInBytes = currentOffset;
+        this.allocator = allocator;
+    }
+
+    @Override
+    public long size() {
+        return sizeInBytes;
+    }
+
+    private int getStartBufferIndexForOffset(long offset) {
+        // It is super common to read from the start or end of a file (the header or footer)
+        // so special case them
+        if (offset == 0) {
+            return 0;
+        }
+        int startIndex = 0;
+        int endIndex = startOffsets.length - 1;
+        if (offset >= startOffsets[endIndex]) {
+            return endIndex;
+        }
+        while (startIndex != endIndex) {
+            int midIndex = (int)(((long)startIndex + endIndex) / 2);
+            long midStartOffset = startOffsets[midIndex];
+            if (offset >= midStartOffset) {
+                // It is either in mid or after mid.
+                long midEndOffset = Long.MAX_VALUE;
+                if (midIndex < startOffsets.length - 1) {
+                    midEndOffset = startOffsets[midIndex + 1];
+                }
+                if (offset <= midEndOffset) {
+                    // We found it in mid
+                    return midIndex;
+                } else {
+                    // It is after mid
+                    startIndex = midIndex + 1;
+                }
+            } else {
+                // It is before mid
+                endIndex = midIndex - 1;
+            }
+        }
+        return startIndex;
+    }
+
+
+    interface DoCopy<T extends MemoryBuffer> {
+        void copyFromHostBuffer(T dest, long destOffset, HostMemoryBuffer src,
+                                long srcOffset, long srcAmount);
+    }
+
+    private <T extends MemoryBuffer> long read(long offset, long amount,
+                                               T dest, DoCopy<T> doCopy) {
+        long realOffset = Math.min(offset, sizeInBytes);
+        long realAmount = Math.min(sizeInBytes - realOffset, amount);
```

Comment:
But then why does the CUDF API bother to return the amount read at all? That feels like too much of a departure from the CUDF APIs for me.

Reply:
OK that makes sense, but what about negative values which the cudf C++ interface does not allow since it's unsigned values?  Pretty sure we're going to do some bad and hard to debug things if someone accidentally passes a negative offset or length to this.